mlir/lib/Dialect/Ptr/IR/PtrDialect.cpp

```diff
@@ -39,6 +41,37 @@ void PtrDialect::initialize() {
       >();
 }
 
+//===----------------------------------------------------------------------===//
+// PtrAddOp
+//===----------------------------------------------------------------------===//
+
+/// Fold the op to the base ptr when the offset is 0.
+OpFoldResult PtrAddOp::fold(FoldAdaptor adaptor) {
+  Attribute attr = adaptor.getOffset();
+  if (!attr)
+    return nullptr;
+  if (llvm::APInt value; m_ConstantInt(&value).match(attr) && value.isZero())
+    return getBase();
+  return nullptr;
+}
+
+Value PtrAddOp::getViewSource() { return getBase(); }
+
+//===----------------------------------------------------------------------===//
+// TypeOffsetOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult TypeOffsetOp::fold(FoldAdaptor adaptor) {
+  return TypeAttr::get(getElementType());
```

Comment:
I don't quite get what this is?

Reply:
`ConstantLike` requires to always return an attribute. In the last commit I added the `type_offset` attribute to make it unambiguous.

Reply:
I'm confused by what you're trying to achieve with this: why is this a "ConstantLike" in the first place?
I don't think we should fold to anything else than an attribute matching the type of the SSA value.

Reply:
Right now, I'm folding it with `TypeOffsetAttr getOffset()`, a int/index typed attribute representing the unrealized constant value. Also, we cannot fold it to an `IntegerAttr` if there's no DL.

Now, the op is `ConstantLike`, because it represents a constant with respect to a potential DL. Also, if it's not constant then:
```mlir
%0 = ptr.type_offset <f32> : i32
%1 = ptr.type_offset <f32> : i32
```
Never gets simplified to `%0 = %1`.

Reply:
I don't quite get it, just being "Pure" should be enough for CSE to merge:
```
%0 = ptr.type_offset <f32> : i32
%1 = ptr.type_offset <f32> : i32
```

Reply:
Yes, CSE does the trick. I was thinking of that op as a constant, thus we get equivalence during canonicalization without having to invoke another transformation.

I'll change it.